src/arch/src/aarch64/regs.rs

```diff
@@ -196,6 +207,236 @@ pub fn read_mpidr(vcpu: &VcpuFd) -> Result<u64> {
+    // First one, stack pointer.
+    let off = offset__of!(user_pt_regs, sp);
+    state.regs.sp = vcpu
+        .get_one_reg(arm64_core_reg_id!(KVM_REG_SIZE_U64, off))
+        .map_err(Error::GetCoreRegister)?;
+
+    // Second one, the program counter.
+    let off = offset__of!(user_pt_regs, pc);
+    state.regs.pc = vcpu
+        .get_one_reg(arm64_core_reg_id!(KVM_REG_SIZE_U64, off))
+        .map_err(Error::GetCoreRegister)?;
+
+    // Next is the processor state.
+    let off = offset__of!(user_pt_regs, pstate);
+    state.regs.pstate = vcpu
+        .get_one_reg(arm64_core_reg_id!(KVM_REG_SIZE_U64, off))
+        .map_err(Error::GetCoreRegister)?;
+
+    // The stack pointer associated with EL1
+    let off = offset__of!(kvm_regs, sp_el1);
+    state.sp_el1 = vcpu
+        .get_one_reg(arm64_core_reg_id!(KVM_REG_SIZE_U64, off))
+        .map_err(Error::GetCoreRegister)?;
+
+    // Exception Link Register for EL1, when taking an exception to EL1, this register
+    // holds the address to which to return afterwards.
+    let off = offset__of!(kvm_regs, elr_el1);
+    state.elr_el1 = vcpu
+        .get_one_reg(arm64_core_reg_id!(KVM_REG_SIZE_U64, off))
+        .map_err(Error::GetCoreRegister)?;
+
+    // Saved Program Status Registers, there are 5 of them used in the kernel.
+    let mut off = offset__of!(kvm_regs, spsr);
+    for i in 0..KVM_NR_SPSR as usize {
+        state.spsr[i] = vcpu
+            .get_one_reg(arm64_core_reg_id!(KVM_REG_SIZE_U64, off))
+            .map_err(Error::GetCoreRegister)?;
+        off += std::mem::size_of::<u64>();
+    }
+
+    // Now moving on to floting point registers which are stored in the user_fpsimd_state in the kernel:
+    // https://elixir.free-electrons.com/linux/v4.9.62/source/arch/arm64/include/uapi/asm/kvm.h#L53
+    let mut off = offset__of!(kvm_regs, fp_regs) + offset__of!(user_fpsimd_state, vregs);
+    for i in 0..32 {
+        state.fp_regs.vregs[i][0] = vcpu
+            .get_one_reg(arm64_core_reg_id!(KVM_REG_SIZE_U128, off))
+            .map_err(Error::GetCoreRegister)?;
+        off += mem::size_of::<u128>();
+    }
+
+    // Floating-point Status Register
+    let off = offset__of!(kvm_regs, fp_regs) + offset__of!(user_fpsimd_state, fpsr);
+    state.fp_regs.fpsr = vcpu
+        .get_one_reg(arm64_core_reg_id!(KVM_REG_SIZE_U32, off))
+        .map_err(Error::GetCoreRegister)? as u32;
+
+    // Floating-point Control Register
+    let off = offset__of!(kvm_regs, fp_regs) + offset__of!(user_fpsimd_state, fpcr);
+    state.fp_regs.fpcr = vcpu
+        .get_one_reg(arm64_core_reg_id!(KVM_REG_SIZE_U32, off))
+        .map_err(Error::GetCoreRegister)? as u32;
+    Ok(())
+}
+
+/// Save the state of the core registers.
+///
+/// # Arguments
+///
+/// * `vcpu` - Structure for the VCPU that holds the VCPU's fd.
+/// * `state` - Structure containing the state of the core registers.
+pub fn restore_core_registers(vcpu: &VcpuFd, state: &kvm_bindings::kvm_regs) -> Result<()> {
```

Comment:
I did think about this but I did not like what I obtained. At every step you need to check for every register if  you are in a get or a set since the offset is calculated progressively and that makes the function uglier than this. I d also need to use a function which receives mutable objects just for the "get" usecase and ignore the mutable for the set. I find this alternative harder to read, wdyt?